### PR TITLE
feat: support to reject write after flushing

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Support to reject write after flushing

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new target `fuzz_migrate_mito_regions` with a "Local WAL" mode for enhanced configuration.
	- Added a `reject_write` field to the `DowngradeRegion` struct for better management of write requests.

- **Bug Fixes**
	- Updated logic to set region roles more explicitly rather than managing writable states.

- **Documentation**
	- Improved documentation comments for clarity, including updates to method descriptions.

- **Refactor**
	- Renamed `leader_status` to `leader_state` across multiple components for consistency and clarity.

### CODE REVIEW

**Comment by WenyXu** [comment]
@coderabbitai summary

**Comment by killme2008** [comment]
@WenyXu Lots of conflicts.

**Comment by WenyXu** [comment]
> @WenyXu Lots of conflicts.

Interesting, I'm going to fix it 🥲

File: src/mito2/src/engine/basic_test.rs [hunk 1]
```diff
@@ -88,6 +88,41 @@ async fn test_write_to_region() {
...
+    let region_id = RegionId::new(1, 1);
+    let request = CreateRequestBuilder::new().build();
+
+    let column_schemas = rows_schema(&request);
+    engine
+        .handle_request(region_id, RegionRequest::Create(request))
+        .await
+        .unwrap();
+    let result = engine
+        .set_region_role_state_gracefully(region_id, SettableRegionRoleState::DowngradingLeader)
```
Comment (evenyag) [inline src/mito2/src/engine/basic_test.rs:105]: We should also add a case that puts some rows before downgrading.

File: src/mito2/src/worker.rs [hunk 2]
```diff
@@ -834,22 +839,23 @@ impl<S: LogStore> RegionWorkerLoop<S> {
...
     ) {
         if let Some(region) = self.regions.get_region(region_id) {
             // We need to do this in background as we need the manifest lock.
             common_runtime::spawn_global(async move {
-                region.set_readonly_gracefully().await;
+                region.set_role_state_gracefully(region_role_state).await;
 
                 let last_entry_id = region.version_control.current().last_entry_id;
-                let _ = sender.send(SetReadonlyResponse::success(Some(last_entry_id)));
+                let _ = sender.send(SetRegionRoleStateResponse::success(Some(last_entry_id)));
```
Comment (evenyag) [inline src/mito2/src/worker.rs:855]: Is it possible that the region handles new write requests before the background task setting the region state? Can we ensure the integrity of the data?

File: src/store-api/src/region_engine.rs [hunk 3]
```diff
@@ -85,19 +109,26 @@ impl From<PbGrantedRegion> for GrantedRegion {
...
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
 pub enum RegionRole {
     // Readonly region(mito2)
     Follower,
     // Writable region(mito2), Readonly region(file).
     Leader,
+    // Leader is downgrading to follower.
+    //
+    // This state is used to prevent new write requests.
+    DowngradingLeader = 2,
```
Comment (evenyag) [inline src/store-api/src/region_engine.rs:123]: Why do we assign a number to this?

File: src/mito2/src/flush.rs [hunk 4]
```diff
@@ -409,9 +409,17 @@ impl RegionFlushTask {
         let action_list = RegionMetaActionList::with_action(RegionMetaAction::Edit(edit.clone()));
         // We will leak files if the manifest update fails, but we ignore them for simplicity. We can
         // add a cleanup job to remove them later.
-        self.manifest_ctx
-            .update_manifest(RegionState::Writable, action_list)
+        let version = self
+            .manifest_ctx
+            .update_manifest(
+                smallvec![RegionLeaderState::Writable, RegionLeaderState::Downgrading],
```
Comment (evenyag) [inline src/mito2/src/flush.rs:415]: Looks like only this step needs a vec of `RegionLearderState`. I think we can add a `FlushReason` variant for downgrading so we can pass `RegionLeaderState::Downgrading` to the manifest context.

Then `update_manifest()` doesn't need to take a smallvec as input. It also ensures the state is downgrading.

File: src/mito2/src/region.rs [hunk 5]
```diff
@@ -376,18 +399,92 @@ impl ManifestContext {
...
-        manager.update(action_list).await.inspect_err(
+        let version = manager.update(action_list).await.inspect_err(
             |e| error!(e; "Failed to update manifest, region_id: {}", manifest.metadata.region_id),
         )?;
 
-        if self.state.load() == RegionState::ReadOnly {
+        if self.state.load() == RegionRoleState::Follower {
             warn!(
-                "Region {} becomes read-only while updating manifest which may cause inconsistency",
+                "Region {} becomes follower while updating manifest which may cause inconsistency",
```
Comment (evenyag) [inline src/mito2/src/region.rs:408]: Also logs the version?

File: src/mito2/src/region.rs [hunk 6]
```diff
@@ -376,18 +399,92 @@ impl ManifestContext {
...
+    ///
+    /// Transition:
+    /// - Follower -> Leader
+    /// - Downgrading Leader -> Leader
+    /// - Leader -> Follower
+    /// - Downgrading Leader -> Follower
+    /// - Leader -> Downgrading Leader
+    ///
+    /// ```
+    pub(crate) fn set_role_role_state(&self, next_role: RegionRole, region_id: RegionId) {
```
Comment (evenyag) [inline src/mito2/src/region.rs:437]: Can we rename it to `set_role()` or `try_set_role()`? We only pass `RegionRole` to it.

File: src/mito2/src/worker.rs (hunk 2 above)
Comment (WenyXu) [reply src/mito2/src/worker.rs:855]: Yes, in the upper layer. we will convert region to downgrading first, and call flush memtable, finally convert it to readonly.

File: src/store-api/src/region_engine.rs (hunk 3 above)
Comment (WenyXu) [reply src/store-api/src/region_engine.rs:123]: AI did🥲

File: src/mito2/src/flush.rs (hunk 4 above)
Comment (WenyXu) [reply src/mito2/src/flush.rs:415]: Good point

File: src/common/meta/src/rpc/router.rs [hunk 7]
```diff
@@ -269,49 +273,50 @@ pub struct RegionRoute {
...
-    /// The following cases in which the [Region] will be downgraded.
+pub enum RegionState {
+    /// The following cases in which the [`Region`] will be downgraded.
     ///
-    /// - The [Region] is unavailable(e.g., Crashed, Network disconnected).
-    /// - The [Region] was planned to migrate to another [Peer].
-    Downgraded,
+    /// - The [`Region`] may be unavailable (e.g., Crashed, Network disconnected).
+    /// - The [`Region`] was planned to migrate to another [`Peer`].
+    Downgrading,
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/rpc/router.rs:292]: The downgraded leader equals a follower

File: src/datanode/src/heartbeat/handler/downgrade_region.rs [hunk 8]
```diff
@@ -66,17 +72,44 @@ impl HandlerContext {
...
+                return self.become_follower_gracefully(region_id).await;
             }
 
             let region_server_moved = self.region_server.clone();
             if let Some(flush_timeout) = flush_timeout {
+                if reject_write {
+                    // Sets region to downgrading, the downgrading region will reject all write requests.
+                    match self
+                        .region_server
+                        .downgrade_region_gracefully(region_id)
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/datanode/src/heartbeat/handler/downgrade_region.rs:84]: we can ensure the region is downgrading after calling this.

File: src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs [hunk 9]
```diff
@@ -95,15 +97,34 @@ impl DowngradeLeaderRegion {
...
+            .get(&region_id.region_number())
+        {
+            let options: WalOptions = serde_json::from_str(wal_option)
+                .with_context(|_| error::DeserializeFromJsonSnafu { input: wal_option })?;
+            matches!(options, WalOptions::RaftEngine)
+        } else {
+            true
+        };
+
+        Ok(reject_write)
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs:125]: I prefer it this way:

```
    if let Some(wal_option) = datanode_table_value
        .region_info
        .region_wal_options
        .get(&region_id.region_number())
    {
        let options: WalOptions = serde_json::from_str(wal_option)
            .with_context(|_| error::DeserializeFromJsonSnafu { input: wal_option })?;
        return Ok(matches!(options, WalOptions::RaftEngine));
    }

    Ok(true)
```

Comment (fengjiachun) [reply src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs:125]: Or `let Some(was_options) = ...  else { return  OK(true); }`

File: src/mito2/src/flush.rs [hunk 10]
```diff
@@ -407,11 +409,22 @@ impl RegionFlushTask {
...
         // We will leak files if the manifest update fails, but we ignore them for simplicity. We can
         // add a cleanup job to remove them later.
-        self.manifest_ctx
-            .update_manifest(RegionState::Writable, action_list)
+        let version = self
+            .manifest_ctx
+            .update_manifest(expected_state, action_list)
             .await?;
+        info!(
+            "Successfully update manifest version to {version}, region: {}",
```
Comment (evenyag) [inline src/mito2/src/flush.rs:425]: Do we need to log the flush reason?

**Review by evenyag — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
It will be LGTM after when all the conversions are resolved.

**Pull request merged**